pippet resource support

ffrank opened the pull request:
The pippet resource implements faster integration of Puppet resources
in mgmt at runtime, by piping synchronization commands to a Puppet
process that keeps running alongside mgmt. This avoids huge overhead
through launching a Puppet process for each operation on a resource
that is delegated to Puppet.

ffrank commented:
Missing context here: Once this is in, the Puppet->mgmt translator can stop making this:
```
exec "whatever" { cmd => "puppet yamlresource resname restype { params ... }", ... }
```

And instead make this
```
pippet "whatever" { name: "resname", type: "restype", params: "{ <yaml stuff> }" }
```

That way, when the translator has to decide to fall back to Puppet, synchronizing the graph will not be as atrociously slow.

purpleidea commented:
Should I be reviewing or is this still meant as WIP?

ffrank commented:
@purpleidea this is now ready for its first review.

The PuppetReceiver code (in util) lacks logging, which is unfortunate. Should probably be added before merge, but I'm not yet sure how to approach it.

Another sore spot are missing tests. We'd need to add Puppet to the build requirements. This is necessary anyway, I feel, in order to make sure that any Puppet integration/support works.

purpleidea commented:
@ffrank Cool, before I review can you move: `util/pippet_run.go` under resources? If you *need* a new package, you can do so below that dir like packagekit does. (Probably not necessary here.) Util is for stuff that's common to more than one resource or package.

ffrank commented:
I had tried that originally, but failed to get it to work because of import cycles. I'll see if I can resolve this but it might not work out.

ffrank commented:
Never mind. It's done.

ffrank commented:
Addressed the warnings I could manage.
I don't understand why the gofmt test fails, it's what I use on my end. Wrong go version? (This is terrible then.)

purpleidea commented:
> gofmt test fails

Yeah, it's very shitty, but they break compat between versions. At least twice so far I think. Unfortunately we'll have to use 1.10 until Fedora 30 is out, and then I'll move to 1.11.

Reviewing now...

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,183 @@
+
+import (
+	"fmt"
+	//"sync"
```
rm please

ffrank replied:
Done.

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,183 @@
+	engine.RegisterResource("pippet", func() engine.Res { return &PippetRes{} })
+}
+
+// PippetRes is a wrapper resource for puppet.
```
Wrap and reflow everything to 80 chars please.

ffrank replied:
Fixed.

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,183 @@
+
+	init *engine.Init
+
+	Type   string `yaml:"type" json:"type"`
```
Why json?
Also, you'll want to add the `lang` tags only if you want anyone to be able to use this from the DSL. (Do we?)

ffrank replied:
Oops, sorry. That was from an earlier test I did. Removed.

I would not expose this resource type to the language until we get a specific feature request for this. It should be considered an integral part of the Puppet Crutch.

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,183 @@
+// Default returns an example Pippet resource.
+func (obj *PippetRes) Default() engine.Res {
+	return &PippetRes{
+		Params: "{}",
```
Could you add a comment on this line explaining what this means please?

ffrank replied:
Done.

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,183 @@
+	init *engine.Init
+
+	Type   string `yaml:"type" json:"type"`
+	Title  string `yaml:"title" json:"title"`
```
Each public field can have a docstring comment, eg: https://github.com/purpleidea/mgmt/blob/master/engine/resources/file.go#L63

ffrank replied:
Done.

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,183 @@
+		return err
+	}
+
+	obj.runner.Register(obj)
```
Should this method have an error returned that we should check, or is it guaranteed to always pass?

ffrank replied:
No. You're right. Fixed.

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,183 @@
+
+// Close is run by the engine to clean up after the resource is done.
+func (obj *PippetRes) Close() error {
+	obj.runner.Unregister(obj)
```
if err := ?

ffrank replied:
I went for `return obj.runner.Unregister(obj)`. Is that fine?

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,183 @@
+// CheckApply synchronizes the resource if required.
+func (obj *PippetRes) CheckApply(apply bool) (checkOK bool, err error) {
+	if obj.init.Refresh() {
+		obj.init.Logf("WARNING - pippet does not support refreshing yet")
```
Then just remove this entirely please...

ffrank replied:
Done.

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	if obj.Type != res.Type {
+		return fmt.Errorf("the resource types differ")
```
nit: just because the word `type` is so overloaded, perhaps: `the Type param differs` instead?

ffrank replied:
Done.

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,183 @@
+		return fmt.Errorf("the resource titles differ")
+	}
+
+	// FIXME: This is a lie. Parameter lists can be equivalent
```
You're allowed to do a more complex Cmp if it's appropriate. Is this what you mean? Eg: compare the sorted lists that you mapped with Trim?

ffrank replied:
I don't think it's that simple. Given two arbitrary YAML hashes, you will be hard pressed to make a call on their being equivalent without actually parsing the YAML. And Go cannot casually do this.

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,183 @@
+}
+
+// PippetUID is the UID struct for PippetRes.
+// FIXME: Is it fine to ignore parameters here? S.a. the note in Cmp.
```
S.a ???

ffrank replied:
Ended up removing this note.
Upon further thought, it's fine to use type/title as the UID, because these resources are expected to be generated from Puppet code, which will not allow this combination to be duplicated, anyway.

purpleidea commented on engine/resources/pippet_run.go:
```diff
@@ -0,0 +1,160 @@
+// PippetRunner is the interface to the wrapped "puppet yamlresource receive"
+// process that keeps running alongside mgmt. Individual resources are expected
+// to register themselves, and can use the Apply function from that point on.
+type PippetRunner interface {
```
I'm not sure why you have an interface type. Interfaces are useful (in general) when more than one thing implements them, or when you plan a future second implementation. AIUI there's only one here...

ffrank replied:
It will make testing much easier :)

purpleidea replied:
If so, then can you write a small stub test to show how it's useful/easier that way please?

purpleidea commented on engine/resources/pippet_run.go:
```diff
@@ -0,0 +1,160 @@
+type PippetRunner interface {
+	Register(interface{}) error
+	Unregister(interface{}) error
+	Apply(string, string, string) PippetResult
```
Weird return arg, would: (PippetResult, error) be more appropriate?

ffrank replied:
It's a good point. The result from Puppet's sync operation is unmarshalled from JSON, in PippetResult form, and that *has* to include any errors that happened (we have no other way to get diagnostics from Puppet to mgmt).

Of course, we could go ahead and extract the interesting fields within the Apply function, and then return something that needs not be a PippetResult plus an error. Would you prefer that?

purpleidea replied:
I think you want (PippetResult, error) here... Let's see what that looks like.

ffrank replied:
I reworked the function, and it became apparent that returning (bool, error) is sufficient (I feel).

Let me know what you think. Keeping this in a separate commit for now, so you can review the change itself if you wish.

purpleidea commented on engine/resources/pippet_run.go:
```diff
@@ -0,0 +1,160 @@
+// PippetResult is the structured return value type for the PippetRunner's
+// Apply function.
+type PippetResult struct {
+	Error     bool
```
What's the difference between Error and failed? And Exception?

My guess is you might prefer:

`(PippetResult, error)` (as mentioned above), with:

```type PippetResult struct{
   Changed bool
   AnyOtherNeededParam int
}```

ffrank replied:
It's subtle, so I'm not sure we need it.

"Failed" is a flag that puppet can set during the sync operation. E.g., a *mount* resource could not mount the file system.

"Error" is a flag my own module sets to indicate that there was an issue with the input, e.g. a YAML parsing error, or an invalid resource type.

"Exception" is the message that describes what went wrong specifically in case Error==true.

A failed resource does not fill the Exception field. Any diagnostics will be written to stderr by Puppet itself, and we don't bother trying to catch those atm. The result is some red Puppet output mixed into mgmt's own output. This is not future-proof, but I would prefer not to over-engineer this detail right now.

ffrank replied:
I think this is "fine" now because the Apply function was reworked. (?)

purpleidea commented on engine/resources/pippet_run.go:
```diff
@@ -0,0 +1,160 @@
+	registered    int
+}
+
+var pippetRunnerInstance PippetRunner
```
global vars (if required) should be at the top of the file, right after imports. Basically so they're easily seen, since they can be problematic if you aren't expecting them.

ffrank replied:
Fixed.

purpleidea commented on engine/resources/pippet_run.go:
```diff
@@ -0,0 +1,160 @@
+// Init runs the Puppet process that will perform the work of synchronizing resources
+// that are sent to its stdin. The process will keep running until Close is called.
+// Init should not be called directly. It is implicitly called by the Register function.
+func (obj *pippetReceiver) Init() error {
```
method receivers should follow their struct... Keep them all together.

purpleidea commented on engine/resources/pippet_run.go:
```diff
@@ -0,0 +1,160 @@
+	}
+	obj.stdout, err = cmd.StdoutPipe()
+	if err != nil {
+		obj.stdin.Close()
```
err ?

ffrank replied:
FIxed.

purpleidea commented on engine/resources/pippet_run.go:
```diff
@@ -0,0 +1,160 @@
+		return err
+	}
+	if err := cmd.Start(); err != nil {
+		obj.stdin.Close()
```
err?

ffrank replied:
Fixed.

purpleidea commented on engine/resources/pippet_run.go:
```diff
@@ -0,0 +1,160 @@
+// Register should be called by any user (i.e., any pippet resource) before using
+// the Apply function on this receiver object.
+// Register implicitly takes care of calling Init if required.
+func (obj *pippetReceiver) Register(res interface{}) error {
```
The input isn't even used in this function... ?

ffrank replied:
True, it's just flavor now - the resource registers "itself" even though the runner object uses a simple counter.
I removed it.

purpleidea commented on engine/resources/pippet_run.go:
```diff
@@ -0,0 +1,160 @@
+	if err = obj.Init(); err != nil {
+		obj.registered = 0
+	}
+	obj.registerMutex.Unlock()
```
Use defer to Unlock instead.

ffrank replied:
D'oh. So much nicer now. Fixed.

purpleidea commented on engine/resources/pippet_run.go:
```diff
@@ -0,0 +1,160 @@
+// Mgmt
```
I'm not sure that this needs to be in a separate file. Unless you have a good reason, put the contents at the bottom of the pippet.go resource file.

ffrank replied:
Done.

purpleidea commented:
@ffrank Is everything addressed and ready for a new review?

ffrank commented:
I have yet to add a test. The rest is addressed, I believe.

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,325 @@
+}
+
+// PippetRes is a wrapper resource for puppet.  It implements the functional
+// equivalent of an exec resource that calls "pupept resource <type> <title>
```
pupept

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,325 @@
+		return err
+	}
+
+	if err = obj.runner.Register(); err != nil {
```
`err :=` inside a new scope

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,325 @@
+}
+
+// CheckApply synchronizes the resource if required.
+func (obj *PippetRes) CheckApply(apply bool) (checkOK bool, err error) {
```
Get rid of the named return values please to be consistent with https://github.com/purpleidea/mgmt/commit/e4d874cc69837784527ae79044d7c0cd0e7ebdbc

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,325 @@
+
+// GroupCmp returns whether two resources can be grouped together or not.
+func (obj *PippetRes) GroupCmp(r engine.GroupableRes) error {
+	return fmt.Errorf("pippet resources are never grouped")
```
Just omit the method then.

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,325 @@
+	}
+	obj.stdout, err = cmd.StdoutPipe()
+	if err != nil {
+		err = errwrap.Append(err, obj.stdin.Close())
```
In this case you can just `return errwrap.Append(err, obj.stdin.Close())`

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,325 @@
+		return err
+	}
+	if err = cmd.Start(); err != nil {
+		err = errwrap.Append(err, obj.stdin.Close())
```
same as above

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,325 @@
+		return err
+	}
+	buf := make([]byte, 80)
+	if _, err = obj.stdout.Read(buf); err != nil {
```
errwrap allows you to inline these...

iow:

```
err = obj.stdout.Read(buf)
err = errwrap.Append(err, obj.stdin.Close())
```

IIRC

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,325 @@
+	obj.applyMutex.Lock()
+	defer obj.applyMutex.Unlock()
+	if err = json.NewEncoder(obj.stdin).Encode(resource); err != nil {
+		return false, fmt.Errorf("failed to send resource to puppet: %v", err)
```
errwrap.Wrapf instead

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,325 @@
+		err = obj.Close()
+	}
+	if obj.registered < 0 {
+		err = errwrap.Append(err, fmt.Errorf("pippet runner: ERROR: unregistered more resources than were registered"))
```
s/....

```
e := fmt.Errorf("pippet runner: unregistered more resources than were registered")
err = errwrap.Append(err, e)
```

purpleidea commented:
(Also, please rebase to master.)

ffrank commented:
Heyyyy, so, this sat for way too long. I tuned out of it for a while there, mulling around mocks and tests in my head. I think at one point you had challenged me to show why I wanted that interface in there so badly? I don't clearly recall, and cannot find the comment anymore.

So I refactored a bit so that I can now use Dependency Injection for testing. I'm rather pleased with it. Let me know your thoughts.

I also rebased and tried to act on all your comments. There is a separate commit up top that has all the changes regarding the review comments.

Thanks for the patience!

purpleidea commented:
Ping when the basic tests pass and I'll review. There is one race that occasionally causes the shell tests to fail. You can ignore that. (It's in the middle group.)

ffrank commented:
@purpleidea this is ready for another round of review. Thanks in advance.

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,335 @@
+	engine.RegisterResource("pippet", func() engine.Res { return &PippetRes{} })
+}
+
+// PippetRes is a wrapper resource for puppet.  It implements the functional
```
We only use single spaces. Both here and in word processors =D

purpleidea commented on engine/resources/pippet.go:
```diff
@@ -0,0 +1,335 @@
+
+	// Type is the exact name of the wrapped Puppet resource type, e.g.
+	// "file", "mount". This needs not be a core type. It can be
+	// a type from a module. The Puppet installation local to the mgmt
```
All docstrings *must* be wrapped to 80 chars. This means 80, not less than 80, not more than 80. Please check the entire patch for this. I recommend using an editor that either does this for you and/or adds a line at 80 chars. Sadly we don't have an automatic test for this yet. Please write one =D

ffrank replied:
I may take that challenge some day.

But not today. :)

ffrank commented:
I wrapped the docstrings to the best of my ability.
I will squash all the commits once you feel this is ready for merge.

ffrank commented:
@purpleidea ready for another round whenever you are.

purpleidea commented:
@ffrank Sorry for the delay, as you know I was sick, but I'm back and catching up on life.

If you rebase to git master, the tests should look a lot better now. I fixed the bbolt issue mostly.

There's still one failing thing of importance, but at least the whole thing is failing less and only in that one test mostly.

Also etcd logging changed in the new version so that's super ugly and needs fixing too :( Patches welcome!

ffrank commented:
@purpleidea great, thanks. Pushed the rebase. Tests fail pretty badly indeed, but it does build 👍 Cheers!

purpleidea commented:
Okay, sorry I was AFK for a bit, as I mentioned to you in IRC, I was quite sick :( Anyways, I decided to merge this and get it done. I squashed all your commits. I also fixed all the wrapping issues I saw at a quick glance, and added these (and a few tiny fixups) in a separate commit here: https://github.com/purpleidea/mgmt/commit/d34715b4ba5bfe9f53a3c5826de77b994fdbda33

Please have a look at figure out why your editor doesn't see these. To be fair, it's something we should totally have a test for. At least for the docstring parts. That could be pretty easy to write if we use the golang AST to pull out the docstrings and check that they're wrapped correctly. If you're in the mood, that would be a good patch for you to write!

Cheers!
James
